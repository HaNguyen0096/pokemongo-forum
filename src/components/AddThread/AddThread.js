import React, { Component } from 'react'
import TopicContext from '../../contexts/TopicContext'
import ApiService from '../../services/api-service'
import { Button, Textarea } from '../Utils/Utils'
import './AddThread.css'

export default class AddThread extends Component {
  
  static contextType = TopicContext

  handleSubmit = ev => {
    ev.preventDefault()
    const { topic } = this.context
    console.log(topic)
    const { title, content } = ev.target

    ApiService.postThread(topic.id, title.value, content.value)
      .then(this.context.addThread)
      .catch(this.context.setError)
  }

  render() {
    return (
      <form
        className='Thread'
        onSubmit={this.handleSubmit}
      >
        <div className='addThread'>
          <div className='title'>
          <Textarea
            required
            aria-label='Thread Title...'
            name='title'
            id='title'
            cols='30'
            rows='2'
            placeholder='Thread Title..'>
          </Textarea>
          </div>
          <div className='content'>
          <Textarea
            required
            aria-label='Type your thread...'
            name='content'
            id='content'
            cols='30'
            rows='4'
            placeholder='Type your thread..'>
          </Textarea>
          </div>
        </div>

        <Button type='submit'>
          Post Thread
        </Button>
      </form>
    )
  }
}
