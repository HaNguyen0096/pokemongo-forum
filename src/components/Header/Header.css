*{
    box-sizing: border-box;
}

.headerTop {
    display: block;
}

img{
    background-size: cover;
    width: 100%;
}

.loginNav{
    display: flex;
    align-items: center;
    justify-content: space-between;
    background-color: #aaa;
    font-size: 30px;
}

.title{
    display: fix;
    text-align: left;
    margin-left: 10px;
    font-size: 20px;
}

.login-button{
    margin-right: 10px;
}

.TopicLists{
    display: flex;
    align-items: center;
    justify-content: space-between;
    background-color: #ccc;
    font-size: 30px;
    list-style: none;
}
