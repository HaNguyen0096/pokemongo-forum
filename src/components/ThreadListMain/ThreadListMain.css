.threadList{
  display: block;
  padding: 0;
  margin-left: 10px;
}

.threadBar{

  display: inline-block;
  position: relative;
  font-weight: bold;
  vertical-align: bottom;
  margin-bottom: 10px;
}

.threadTitle{
  text-decoration: none;
}

.threadTitle:link{
  color: black;
}

.threadTitle:visited{
  color: black;
}

.threadTitle:hover{
  color: rgb(23,156,106);
}

.threadContent{
  margin-bottom: 10px;
}

.threadTime{
  color: rgb(113,113,113)
}

.CountLike{
  color: rgb(113,113,113);
  display: flex;
  /* align-items: center; */
  justify-content: flex-start;
}






