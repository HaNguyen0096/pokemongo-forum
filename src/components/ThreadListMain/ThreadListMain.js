import React, { Component} from 'react'
import { Link } from 'react-router-dom'

export default class ThreadListMain extends Component{
  render() {
    const { thread, topic } = this.props
    //console.log(thread)
    //console.log(topic.id)

    return (
      <div className='ThreadList'>
        <Link to={`/topic/${topic.id}/${thread.id}`} className='ThreadListItem'>
        <div className='ThreadListItem__title'>
            <h2>{thread.thread_title}</h2>
        </div>
        </Link>
        <h3>{thread.thread_content}</h3>
      </div>      
    )
  }
}