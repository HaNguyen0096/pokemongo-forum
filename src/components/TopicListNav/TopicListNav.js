import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import './TopicList.css'

export default class TopicListNav extends Component{

  render(){
    const { topic } = this.props
    let count = this.props.countThread
    return (
      <div className='topicNav'>
      <nav role="navigation">
        <ul className='topicNav_List'>
          <div className='topicNameBar'>            
            <Link to={`/topic/${topic.id}`} className='topicNames'>
              <FontAwesomeIcon icon='folder'/> {topic.topic_name}                       
            </Link>
          </div>
          <p className='topicDescription'>{topic.topic_content}</p>
          <p className='countThread'>Threads: {count}</p>
          
        </ul>
      </nav>
      <hr />
      </div>
    )
  }
}