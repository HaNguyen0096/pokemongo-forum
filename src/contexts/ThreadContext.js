import React, {Component} from 'react'

const ThreadContext = React.createContext({
  thread: null,
  comments: [],
  error: null,
  setError: () => {},
  clearError: () => {},
  setThread: () => {},
  clearThread: () => {},
  setComments: () => {},
  addComment: () => {},
  deleteComment: () => {},
  deleteThread: () => {},

})

export default ThreadContext

export class ThreadProvider extends Component {

  state = {
    thread: null,
    error: null,
  };

  setError = error => {
    console.error(error)
    this.setState({ error })
  }

  clearError = () => {
    this.setState({ error: null })
  }

  setThread = thread => {
    this.setState({ thread })
  }

  clearThread = () => {
    this.setThread(null)
    this.setComments([])
  }

  setComments = comments => {
    this.setState({ comments })
  }

  addComment = comment => {
    this.setComments([
      ...this.state.comments,
      comment
    ])
  }

  render(){
    const value = {
      thread: this.state.thread,
      comments: this.state.comments,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      setThread: this.setThread,
      setComments: this.setComments,
      clearThread: this.clearThread,
      addComment: this.addComment,
      deleteComment: this.deleteComment,
      deleteThread: this.deleteThread,
    }
    return (
      <ThreadContext.Provider value={value}>
        {this.props.children}
      </ThreadContext.Provider>
    )
  }
}