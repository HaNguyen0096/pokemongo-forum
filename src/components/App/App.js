import React, { Component } from 'react'
import { Route, Switch } from 'react-router-dom'
import Header from '../Header/Header'
import MainPage from '../../routes/MainPage/MainPage'
import TopicPage from '../../routes/TopicPage/TopicPage'
import ThreadPage from '../../routes/ThreadPage/ThreadPage'
import LoginPage from '../../routes/LoginPage/LoginPage'
import RegistrationPage from '../../routes/RegistrationPage/RegistrationPage'
import NotFoundPage from '../../routes/NotFoundPage/NotFoundPage'
//import ThreadContext from '../../contexts/ThreadContext'
//import config from '../../config'

//import Footer from '../Footer/Footer';
import './App.css'


class App extends Component {

  state = { 
    hasError: false 
  }

  static getDerivedStateFromError(error) {
    console.error(error)
    return { hasError: true }
  }

  render() {
    return(
      < >
        <div className='App'>
        <header className='App__header'>
          <Header />
        </header>
        <main className='App__main'>
        {this.state.hasError && <p className='red'>There was an error! Oh no!</p>}
          <Switch>
            <Route 
              exact
              path={'/'}
              component={MainPage}
            />
            <Route
              exact
              path={'/login'}
              component={LoginPage}
            />
            <Route
              exact
              path={'/register'}
              component={RegistrationPage}
            />
            <Route
              exact
              path={'/topic/:topicId'}
              component={TopicPage}
            />
            <Route
              path={'/topic/:topicId/:threadId'}
              component={ThreadPage}
            />           
            <Route
              component={NotFoundPage}
            />
          </Switch>
        </main>        
      </div>
      </>     
    )
  }
}

export default App;
