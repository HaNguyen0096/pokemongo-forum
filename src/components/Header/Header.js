import React, {Component} from 'react'
import { Link } from 'react-router-dom'
import banner from './go-banner.png'
import TokenService from '../../services/token-service'
import './Header.css'

export default class Header extends Component{
  handleLogoutClick = () => {
  }

  renderLogoutLink() {
    return (
      <div className='Header__logged-in'>
        <Link style={{ textDecoration: 'none' }}
          onClick={this.handleLogoutClick}
          to='/'>
          Logout
        </Link>
      </div>
    )
  }

  renderLoginLink() {
    return (
      <div className='Header__not-logged-in'>
        <Link className='register-button'
          to='/register'>
          Register
        </Link>
        {' - '}
        <Link className='login-button'
          to='/login'>
          Log in
        </Link>
      </div>
    )
  }

  render(){
    
    return (
      <div className='header'>
        <div className='headerTop'>
          <img src={banner} className="banner" alt="Pokemon GO forum" />
          <div className="loginNav">
            <div className="title">Pokemon GO Forum</div>
            <div className='login-button'>
            {TokenService.hasAuthToken()
              ? this.renderLogoutLink()
              : this.renderLoginLink()}
            </div>
          </div>
        </div>
        <nav className='navBar'>
        </nav>
      </div>
    )
  }
}