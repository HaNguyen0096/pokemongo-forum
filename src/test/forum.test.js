import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import App from '../components/App/App'
import { BrowserRouter } from 'react-router-dom'
import { TopicProvider } from '../contexts/TopicContext'
import { TopicListProvider } from '../contexts/TopicListContext'
import { ThreadProvider } from '../contexts/ThreadContext'
import Header from '../components/Header/Header'
import ThreadListMain from '../components/ThreadListMain/ThreadListMain'
import TopicListNav from '../components/TopicListNav/TopicListNav'
import AddThread from '../components/AddThread/AddThread'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

const testTopic = [
  {
    id: 1,
    topic_name: 'General',
    topic_content: 'General discussion about Pokemon Go!.'
  },
]

const testThread = [
  {
    id: 1,
    thread_title: 'events',
    thread_content: 'When is Valentine event start?',
    user_id: 1,
    topic_id: 1,
    modified: '2020-02-22T21:28:32.615Z'
  }
]

it('renders App without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<BrowserRouter>
    <TopicListProvider>
      <TopicProvider>
        <ThreadProvider>
          <App />
        </ThreadProvider>
      </TopicProvider>
    </TopicListProvider>
  </BrowserRouter>, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('Header component', () => {

  it('renders Header without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                      <Header />
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI for Header as expected', () => {
    const tree = renderer
      .create(<BrowserRouter>
                <Header />
              </BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

  it('renders Header by default', () => {
    const wrapper = shallow(<Header />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})

describe('TopicList component', () => {

  it('renders TopicList without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                      <TopicListNav topic={testTopic}/>
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI for TopicList as expected', () => {
    const tree = renderer
      .create(<BrowserRouter>
                <TopicListNav topic={testTopic} />
              </BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

  it('renders TopicList by default', () => {
    const wrapper = shallow(<TopicListNav topic={testTopic} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });

})

describe('ThreadList component', () => {

  it('renders ThreadList without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                      <ThreadListMain topic={testTopic} thread={testThread}/>
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });
})

describe('AddThread component', () => {

  it('renders AddThread without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                      <AddThread />
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI for AddThread as expected', () => {
    const tree = renderer
      .create(<BrowserRouter>
                <AddThread />
              </BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

  it('renders AddThread by default', () => {
    const wrapper = shallow(<AddThread  />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });

})

