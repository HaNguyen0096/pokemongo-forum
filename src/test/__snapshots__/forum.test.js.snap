// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddThread component renders AddThread by default 1`] = `
<form
  className="Thread"
  onSubmit={[Function]}
>
  <div
    className="addThread"
  >
    <div
      className="addTitle"
    >
      <Textarea
        aria-label="Thread Title..."
        cols="30"
        id="title"
        name="title"
        placeholder="Thread Title.."
        required={true}
        rows="2"
      />
    </div>
    <div
      className="addContent"
    >
      <Textarea
        aria-label="Type your thread..."
        cols="30"
        id="content"
        name="content"
        placeholder="Type your thread.."
        required={true}
        rows="4"
      />
    </div>
  </div>
  <Button
    className="postThreadBtn"
    type="submit"
  >
    Post Thread
  </Button>
</form>
`;

exports[`AddThread component renders the UI for AddThread as expected 1`] = `
<form
  className="Thread"
  onSubmit={[Function]}
>
  <div
    className="addThread"
  >
    <div
      className="addTitle"
    >
      <textarea
        aria-label="Thread Title..."
        className="Textarea "
        cols="30"
        id="title"
        name="title"
        placeholder="Thread Title.."
        required={true}
        rows="2"
      />
    </div>
    <div
      className="addContent"
    >
      <textarea
        aria-label="Type your thread..."
        className="Textarea "
        cols="30"
        id="content"
        name="content"
        placeholder="Type your thread.."
        required={true}
        rows="4"
      />
    </div>
  </div>
  <button
    className="Button postThreadBtn"
    type="submit"
  >
    Post Thread
  </button>
</form>
`;

exports[`Header component renders Header by default 1`] = `
<div
  className="header"
>
  <div
    className="headerTop"
  >
    <img
      alt="Pokemon GO forum"
      className="banner"
      src="go-banner.png"
    />
    <div
      className="loginNav"
    >
      <div
        className="title"
      >
        Pokemon GO Forum
      </div>
      <div
        className="login-button"
      >
        <div
          className="Header__not-logged-in"
        >
          <Link
            className="register-button"
            to="/register"
          >
            Register
          </Link>
           - 
          <Link
            className="login-button"
            to="/login"
          >
            Log in
          </Link>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="navBar"
  />
</div>
`;

exports[`Header component renders the UI for Header as expected 1`] = `
<div
  className="header"
>
  <div
    className="headerTop"
  >
    <img
      alt="Pokemon GO forum"
      className="banner"
      src="go-banner.png"
    />
    <div
      className="loginNav"
    >
      <div
        className="title"
      >
        Pokemon GO Forum
      </div>
      <div
        className="login-button"
      >
        <div
          className="Header__not-logged-in"
        >
          <a
            className="register-button"
            href="/register"
            onClick={[Function]}
          >
            Register
          </a>
           - 
          <a
            className="login-button"
            href="/login"
            onClick={[Function]}
          >
            Log in
          </a>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="navBar"
  />
</div>
`;

exports[`Messages component renders the UI for Header as expected 1`] = `
<div
  className="header"
>
  <div
    className="headerTop"
  >
    <img
      alt="Pokemon GO forum"
      className="banner"
      src="go-banner.png"
    />
    <div
      className="loginNav"
    >
      <div
        className="title"
      >
        Pokemon GO Forum
      </div>
      <div
        className="login-button"
      >
        <div
          className="Header__not-logged-in"
        >
          <a
            className="register-button"
            href="/register"
            onClick={[Function]}
          >
            Register
          </a>
           - 
          <a
            className="login-button"
            href="/login"
            onClick={[Function]}
          >
            Log in
          </a>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="navBar"
  />
</div>
`;

exports[`ThreadList component renders ThreadList by default 1`] = `
<div
  className="threadList"
>
  <div
    className="threadBar"
  >
    <Link
      className="threadTitle"
      to="/topic/undefined/undefined"
    />
  </div>
  <div
    className="threadContent"
  />
  <div
    className="threadTime"
  >
    March 22nd 2020, 10:55:59 pm
  </div>
  <div
    className="CountLike"
  >
    <p>
      Comments: 
    </p>
  </div>
  <hr />
</div>
`;

exports[`ThreadList component renders the UI for ThreadList as expected 1`] = `
<div
  className="threadList"
>
  <div
    className="threadBar"
  >
    <a
      className="threadTitle"
      href="/topic/undefined/undefined"
      onClick={[Function]}
    />
  </div>
  <div
    className="threadContent"
  />
  <div
    className="threadTime"
  >
    March 22nd 2020, 10:55:59 pm
  </div>
  <div
    className="CountLike"
  >
    <p>
      Comments: 
    </p>
  </div>
  <hr />
</div>
`;

exports[`ThreadListMain component renders ThreadListMain without crashing 1`] = `
<div
  className="header"
>
  <div
    className="headerTop"
  >
    <img
      alt="Pokemon GO forum"
      className="banner"
      src="go-banner.png"
    />
    <div
      className="loginNav"
    >
      <div
        className="title"
      >
        Pokemon GO Forum
      </div>
      <div
        className="login-button"
      >
        <div
          className="Header__not-logged-in"
        >
          <Link
            className="register-button"
            to="/register"
          >
            Register
          </Link>
           - 
          <Link
            className="login-button"
            to="/login"
          >
            Log in
          </Link>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="navBar"
  />
</div>
`;

exports[`TopicList component renders Header by default 1`] = `
<div
  className="topicNav"
>
  <nav
    role="navigation"
  >
    <ul
      className="topicNav_List"
    >
      <div
        className="topicNameBar"
      >
        <Link
          className="topicNames"
          to="/topic/undefined"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="folder"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
           
        </Link>
      </div>
      <p
        className="topicContent"
      />
      <p
        className="countThread"
      >
        Threads: 
      </p>
    </ul>
  </nav>
  <hr />
</div>
`;

exports[`TopicList component renders TopicList by default 1`] = `
<div
  className="topicNav"
>
  <nav
    role="navigation"
  >
    <ul
      className="topicNav_List"
    >
      <div
        className="topicNameBar"
      >
        <Link
          className="topicNames"
          to="/topic/undefined"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="folder"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
           
        </Link>
      </div>
      <p
        className="topicContent"
      />
      <p
        className="countThread"
      >
        Threads: 
      </p>
    </ul>
  </nav>
  <hr />
</div>
`;

exports[`TopicList component renders the UI for Header as expected 1`] = `
<div
  className="topicNav"
>
  <nav
    role="navigation"
  >
    <ul
      className="topicNav_List"
    >
      <div
        className="topicNameBar"
      >
        <a
          className="topicNames"
          href="/topic/undefined"
          onClick={[Function]}
        >
           
        </a>
      </div>
      <p
        className="topicContent"
      />
      <p
        className="countThread"
      >
        Threads: 
      </p>
    </ul>
  </nav>
  <hr />
</div>
`;

exports[`TopicList component renders the UI for TopicList as expected 1`] = `
<div
  className="topicNav"
>
  <nav
    role="navigation"
  >
    <ul
      className="topicNav_List"
    >
      <div
        className="topicNameBar"
      >
        <a
          className="topicNames"
          href="/topic/undefined"
          onClick={[Function]}
        >
           
        </a>
      </div>
      <p
        className="topicContent"
      />
      <p
        className="countThread"
      >
        Threads: 
      </p>
    </ul>
  </nav>
  <hr />
</div>
`;

exports[`renders the UI as expected 1`] = `
<div
  className="header"
>
  <div
    className="headerTop"
  >
    <img
      alt="Pokemon GO forum"
      className="banner"
      src="go-banner.png"
    />
    <div
      className="loginNav"
    >
      <div
        className="title"
      >
        Pokemon GO Forum
      </div>
      <div
        className="login-button"
      >
        <div
          className="Header__not-logged-in"
        >
          <a
            className="register-button"
            href="/register"
            onClick={[Function]}
          >
            Register
          </a>
           - 
          <a
            className="login-button"
            href="/login"
            onClick={[Function]}
          >
            Log in
          </a>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="navBar"
  />
</div>
`;

exports[`renders the UI for Header as expected 1`] = `
<div
  className="header"
>
  <div
    className="headerTop"
  >
    <img
      alt="Pokemon GO forum"
      className="banner"
      src="go-banner.png"
    />
    <div
      className="loginNav"
    >
      <div
        className="title"
      >
        Pokemon GO Forum
      </div>
      <div
        className="login-button"
      >
        <div
          className="Header__not-logged-in"
        >
          <a
            className="register-button"
            href="/register"
            onClick={[Function]}
          >
            Register
          </a>
           - 
          <a
            className="login-button"
            href="/login"
            onClick={[Function]}
          >
            Log in
          </a>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="navBar"
  />
</div>
`;
